accounts/abi: fix event tupleUnpack errors + cosmetic cleaning

robert-zaremba opened the pull request:
The fix commits contain fixes for two bugs in event unpacker:

1. Not ignoring Indexed arguments when computing data offset 
2. Wrong expected slice/array size
3. Excessive value extraction from the slice element (`.Elem` method) - which caused panic in tests.

The tests commit contains tests for all cases. For convenience I've added new dependency for doing assertions in tests (github.com/stretchr/testify).

Last commit contains cosmetic cleanup.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 5b4992b3ebf84c2311b2bb83ea8c11cc1adbf2f6
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 5b4992b3ebf84c2311b2bb83ea8c11cc1adbf2f6
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

gballet commented on accounts/abi/event_test.go:
```diff
@@ -17,13 +17,45 @@
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/assert"
```
You're pulling in two libraries that produce a big diff (and code) size, and you're the only one using it. I suggest that you create a different PR to start using those two libraries, so that the content of this PR is only about `tupleUnpack`. If the community is interested in using testify, you will get upvotes and we'll then include it. This PR does too many different things and we can't include it whole.

robert-zaremba replied:
OK, will do that.

robert-zaremba replied:
Testify is a really convenient library when doing tests. I will extract it from this pull request and do another one then.

gballet commented on accounts/abi/method.go:
```diff
@@ -192,6 +194,8 @@ func (m Method) String() string {
 }
 
+// Id returns the canonical representation of the method's signature used by the
+// abi definition to identify method name.
```
identify THE method name

robert-zaremba replied:
Updated, thanks.

robert-zaremba commented on accounts/abi/abi.go:
```diff
@@ -85,7 +85,7 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) (err error) {
 		unpack = event
 	} else {
-		return fmt.Errorf("abi: could not locate named method or event.")
+		return fmt.Errorf("abi: could not locate named method or event")
```
This was reported by `golinit`

robert-zaremba commented on accounts/abi/event.go:
```diff
@@ -57,55 +57,57 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
 				}
 			}
-		case reflect.Slice, reflect.Array:
-			if value.Len() < i {
```
The slice / array size should be done once, and at the beginning. Not every time in the loop. Also the error message was wrong, because it could fail before reaching the end (reporting the wrong expected size).

robert-zaremba commented on accounts/abi/method.go:
```diff
@@ -88,9 +88,21 @@ func (method Method) tupleUnpack(v interface{}, output []byte) error {
-		typ   = value.Type()
+		value    = valueOf.Elem()
+		typ      = value.Type()
+		isStruct = true
```
good point. The only additional type which we could support here is map, and this can be easily handled with enum type, which we will define once (performance, checks...)

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 9cb745517e591c8c1cd3d02401976114aa2ac525
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5a20f1626d23fd7fbcb9999fde319bbb1a5f41a1
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

robert-zaremba commented on accounts/abi/event_test.go:
```diff
@@ -54,3 +85,80 @@ func TestEventId(t *testing.T) {
+			[3]byte{'u', 's', 'd'}},
+		jsonEventPledge,
+		"Can unpack Pledge event into slice",
+	}}
```
All the `Slice` based tests were failing because of the `Elem()` call in `tupleUnpack`:

- Can unpack ERC20 Transfer event into slice
- Can unpack Pledge event into slice

robert-zaremba commented:
@gballet Thanks for review. Please look at my answers above.

- I've updated `isStruct` initialization
- in the comment I've pointed out which tests are failing because of `Elem()` call
- removed the testify dependency and updated tests accordingly to that
- I kept the cosmetic update because it's not that big and solve most of the linter warnings. It was easier for me to do it while working on this package.

Please confirm everything is OK, and I can squash the commits before merging the PR.

robert-zaremba commented on accounts/abi/method.go:
```diff
@@ -88,9 +88,21 @@ func (method Method) tupleUnpack(v interface{}, output []byte) error {
+		isStruct = true
 	)
+	switch value.Kind() {
+	case reflect.Struct:
```
done

robert-zaremba commented on accounts/abi/event.go:
```diff
@@ -57,55 +57,57 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
 
-		switch value.Kind() {
-		case reflect.Struct:
+		if isStruct {
```
as explained below - we should check the type before unpacking the arguments in order to validate the array size once and for all. We can do the switch here, but since we already did it it's easier to reuse what we have learned.

robert-zaremba commented:
@gballet I've squashed the review changes. Each commit has  the description explaining the changes (eg the `Elem()` call issue). It's much easier to review individual commits if you want to see the logical changes. As requested the testify dependency has been removed (with commit).

gballet commented on accounts/abi/method.go:
```diff
@@ -116,20 +127,14 @@ func (method Method) tupleUnpack(v interface{}, output []byte) error {
 			v := value.Index(i)
-			if v.Kind() != reflect.Ptr && v.Kind() != reflect.Interface {
-				return fmt.Errorf("abi: cannot unmarshal %v in to %v", v.Type(), reflectValue.Type())
+			if err := requireAssignable(v, reflectValue); err != nil {
```
You're testing your changes in events.go, but not those in methods.go. The code is duplicated, and therefore in the future it could drift apart. So you need to cover that too.

gballet commented on accounts/abi/event_test.go:
```diff
@@ -54,3 +85,80 @@ func TestEventId(t *testing.T) {
+		expected interface{}
+		jsonLog  []byte
+		name     string
+	}{{
```
Forgot to ask in my previous review: are you going to test for errors as well? Like, the array boundary check that you made some changes for?

robert-zaremba commented:
@gballet @karalabe  I've updated the repo according to the review:

+ Added tests for Method slice unpack
+ researched the use-case for `Elem`: shortly it is there to be able to create a slice with elements we want to unpack to. Elem requires that we provide a slice of pointers into unpack. In my events test-cases I assumed that we we want to unpack directly into slice values. I brought back the `Elem` call and updated tests.

robert-zaremba commented:
Also, since `testify` is merged I've updated (only) new tests to take advantage of it, which saved around 2 lines of code for each check :)

robert-zaremba commented:
Into the cleanning commit I've added one more `golint` complaint fix: unified `Method` method receiver name:  [Receiver names should be consistent across a type's methods. ](https://talks.golang.org/2014/names.slide#11)

robert-zaremba commented on accounts/abi/method.go:
```diff
@@ -24,7 +24,7 @@ import (
 )
 
-// Callable method given a `Name` and whether the method is a constant.
+// Method represents a method ABI entry.
```
Not really. Look at the whole comment. I think it's more clear now. The `Name` is part of it's public interface, and other part of the comment mentions about constant part.

VoR0220 replied:
No it definitely drops critical information regarding the ABI as the field "constant" is a core component of the current ABI. This will eventually be changed to "stateMutability" but for now it is needed, especially if the goal is to reduce breaking changes.

robert-zaremba replied:
Well, the second line tells:
```
If the method is `Const` no transaction needs to be created for this
```
I don't want to insist here. I will update to:
```
Method represents a callable given a `Name` and whether the method is a constant.
```

VoR0220 commented on accounts/abi/event.go:
```diff
@@ -59,53 +59,47 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
-	j := 0
-	for i := 0; i < len(e.Inputs); i++ {
-		input := e.Inputs[i]
+	i, j := -1, 0
```
why is i -1 here? Is there any way to get a comment of some kind clarifying the logic here?

robert-zaremba replied:
Because we are incrementing at the beginning of the of the loop (if the argument is not Indexed) rather then at the end.

VoR0220 replied:
That doesn't seem to clarify things for me. So in the case of an event a(uint b, string indexed c, uint d), how does the -1 help read through b? Again, the logic could use some documentation is all I'm saying.

VoR0220 commented on accounts/abi/event_test.go:
```diff
@@ -17,13 +17,50 @@
+      "indexed": false, "name": "currency", "type": "bytes3"
+  }],
+  "name": "Pledge",
+  "type": "event"
```
Needs a test to take into account static array types.

robert-zaremba replied:
Done

VoR0220 commented on accounts/abi/reflect.go:
```diff
@@ -85,3 +85,12 @@ func set(dst, src reflect.Value, output Argument) error {
 }
+
+// requireAssignable verifies that `dest` and `src` types are compatible
+// and it's possible to safely assign `src` value to `dest`.
```
This does not appear to be what it's actually doing.

robert-zaremba replied:
Right, there is better way to document it. Howe about:

// requireAssignable verifies that `dest` type is assignable.

VoR0220 replied:
That works better.

robert-zaremba replied:
changed to:
```
// requireAssignable assures that `dest` is a pointer and it's not an interface.
```

VoR0220 commented on accounts/abi/method.go:
```diff
@@ -90,54 +90,51 @@ func (method Method) tupleUnpack(v interface{}, output []byte) error {
-	for i := 0; i < len(method.Outputs); i++ {
-		toUnpack := method.Outputs[i]
-		if toUnpack.Type.T == ArrayTy {
+	for i, o := range m.Outputs {
```
can we please use a more descriptive name than `o`?

robert-zaremba replied:
This variable is used very often, and doesn't have a big context. But If you want I can change to something else.

VoR0220 commented on accounts/abi/event.go:
```diff
@@ -59,53 +59,47 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
 			// need to move this up because they read sequentially
 			j += input.Type.Size
 		}
+		i++
```
Okay, I see the logic now, slightly...but this definitely needs documentation because currently it's a bit confusing as to why it's set up this way.

robert-zaremba replied:
OK, will add some comments.

robert-zaremba commented:
@VoR0220  how should I ship this small updates (additianl documentation)? Separate commit or squashing into old ones?

VoR0220 commented:
squash docs into old commit and then throw tests for static arrays on top.

robert-zaremba commented:
Docs squashed, new tests added.

VoR0220 commented on accounts/abi/event_test.go:
```diff
@@ -155,6 +155,16 @@ func TestEventTupleUnpack(t *testing.T) {
+			&[3]byte{'u', 's', 'd'}},
+		jsonEventPledge,
+		"",
+		"Can unpack Pledge event into an array",
```
I think I wasn't clear. This is a good test to have but it isn't exactly what I meant when I said add a test for a static array. Relook at what you have written and consider the following problem.

Let's assume we create an event of the signature `event A(uint[3] indexed a, uint[3] b, string c)`. Currently it is my belief that your current implementation will fail this test and overflow.

VoR0220 commented:
see my comment on the commit.

robert-zaremba commented:
@VoR0220 - you are right, there is a bug in deserializing static array arguments.
But I've added one commit with test for that.

But it's not related to this changeset (it also fails in the parent commit). Since this PR is already long, how about following:

+ removing the static array argument test
+ merging 
+ doing separate PR for static array argument test & fix.

VoR0220 commented:
@robert-zaremba Take a look at this PR, I think both of you are trying to fix the same things, perhaps you could just add your tests to this PR and see if it passes? 

https://github.com/ethereum/go-ethereum/pull/15568

robert-zaremba commented:
Yes, adding the following change from #15568 solves the static array problem:
```		if input.Type.T == ArrayTy {
			// combined index ('i' + 'j') need to be adjusted only by size of array, thus
			// we need to decrement 'j' because 'i' was incremented
			j += input.Type.Size - 1
		}
```

However there is a difference my PR and the other one conflict in a way of handling Indexed arguments (mine ignores them, the #15568 updates the accounting variable: `i`).

Also this PR fixes the event unpack into slice (`accounts/abi: fix event unpack into slice` commit) and solves the linting issues.

I'm adding this change as a new commit (though we can squash it into the static array test).

Probably the easiest and a fair way would be to allow @dshulyak do the above change in this PR - in that case I will remove the latest commit.

VoR0220 commented:
@dshulyak any issues?

dshulyak commented:
no issues, i will make separate pr with that bugfix or include it into this one https://github.com/ethereum/go-ethereum/pull/15618

robert-zaremba commented:
Great! I'm happy that we are finally moving forward. So

+ I'm removing last commit from this Pull Request (the static array argument fix). For reference it's available in the [fix-static-array-event-argument](https://github.com/robert-zaremba/go-ethereum/tree/fix-static-array-event-argument) branch
+ @dshulyak please add your change to this Pull Request (`master` branch in my repo) - I've just added you as a collaborator.
+ Should I remove my static-array test (it will fail without the last [commit](https://github.com/robert-zaremba/go-ethereum/commit/293f550fb98dbc7c03f03524cc7de1a5237de6f1) I've just removed from this PR).

Or please advice if you want to do differently?

dshulyak commented:
@robert-zaremba i would prefer to merge both PRs separately (my change is this one https://github.com/ethereum/go-ethereum/pull/15618), doesn't matter for me in which order.
and i think you can remove a test for array tuple unpack, cause it is not related to the original bugfix.

robert-zaremba commented:
@dshulyak OK. Removed. 
Just in case, the ported fix + test is in [fix-static-array-event-argument](https://github.com/robert-zaremba/go-ethereum/tree/fix-static-array-event-argument) branch.

VoR0220 commented:
@gballet @holiman @fjl @karalabe Can we get these two PRs merged up?

robert-zaremba commented:
Ping. Any update here?

holiman commented:
Did you just rebase this ? 
I was just doing that locally aswell :)

gballet commented:
yes I did, there seems to be a merge error though

robert-zaremba commented:
Let me know if you need some help with rebase or squashe. I didn't squash last commits because they were after the approved review.

gballet commented:
that should be fine now, if things pass then we're good to merge

robert-zaremba commented:
It seams that there is an unrelated dependency build (installation) problem for GETH_ARCH=386, MSYS2_ARCH=i686, MSYS2_BITS=32, MSYSTEM=MINGW32

holiman commented:
Closing due to #15731
